namespace Test is
    use System.String;
    use System.Exception;
    use System.Console;

    class Main is
        entry() static is
            @IL.entrypoint()

            try
                for i in 0..10 do
                    try
                        Console.write_line("outer: " + i);
    
                        for j in 0..10 do
                            Console.write_line("inner: " + j);

                            try
                                if i == j then
                                    Console.write_line("inner: " + i + " == " + j + " will throw Exception");
                                    throw new Exception("Exception: " + i + " == " + j);
                                fi
                                

                                if j == 2 then
                                    Console.write_line("inner: " + j + " == 2 will throw MyException");
                                    throw new MyException("MyException: " + j + " == 2");
                                fi
                                
                                if j >= 4 && j <= 5 then
                                    Console.write_line("inner: will continue");
                                    continue;
                                fi                                

                                if j >= 7 then
                                    Console.write_line("inner: will break on 7");
                                    break;
                                fi
                            catch ex: MyException
                                Console.write_line("inner: caught: " + ex.message);                                
                            yrt

                            Console.write_line("inner end");
                        od
    
                    catch ex: Exception
                        Console.write_line("outer: caught: " + ex.message);

                    yrt                
                od
                    

            catch ex: Exception
                Console.write_line("main: caught: " + ex.message);

            yrt
        si
    si

    class MyException: Exception is
        init(message: String) is super.init(message); si
    si   
si